client/deployment_create.go

```diff
@@ -20,6 +20,11 @@ type DeploymentFile struct {
 	Sha  string `json:"sha"`
 	Size int    `json:"size"`
 }
+type GitSource struct {
+	RepoId string `json:"repoId"`
```

Comment:
Digging into this, where does the `repoId` come from? 
I can see in the example that you have it as a number. 🤔. I wonder if we need some data source to expose this so it's more usable?

Reply:
Okay, so `repoId` looks like it is exposed through a project. I think it could be exposed from both the `vercel_project` resource and the `vercel_project` data source. 
This would mean you can use it like: 
```terraform
data "vercel_project" "my_project" {
   name = "my-project"
}

resource "vercel_deployment" "example" {
  team_id = var.vercel_team_id
  project_id = data.vercel_project.my_project.id
  git_source = {
    ref = "main" // or any other branch you fancy. 
    repo_id = data.vercel_project.my_project.git_repository.repo_id
    type = data.vercel_project.my_project.git_repository.type
  }
}
```

I think this would be more preferable than having to dig out a repoId. What do you think @WCMinor?

Reply:
That's true that you can find it out that way, but I think it's somewhat surprising to see (and results in a magic number). I think if we had the data source, or exposed it via a project it would be more intuitive on how to use the two together to trigger a deployment.

Reply:
true, should I create an issue for that?

Reply:
Okay, so the create-deployment API looks a tad more complicated as it supports a few variations of `type (github)/ref/repoId` or `type (github)/ref/org/repo` or `type (gitlab) /ref/projectId` or `type (bitbucket)/ref/repoUuid` or `type (bitbucket)/ref/owner/slug`.

Reply:
I'll give this some thought and try to get back to you earlier next week.

Reply:
I'm curious on the need part - how do you currently deploy (prior to terraform)?

Reply:
we build and install the provider on the deployment ci pipeline and change the Terraform configuration to use the local plugin instead of the public provider. Not ideal.

Reply:
prior to terraform we used vercel cli directly on github actions

Reply:
then we decided to implement a deployment using the vercel API directly, which quickly turned into just fork and use your provider which already does 90% of what we need. BTW we will need in the future to make dns aliases for deployments as well, not sure if this is already implemented in the provider, happy to help there too

Reply:
I have some work in progress for DNS records, assuming this is what you mean? The battle on my part is just getting time to do everything I intend to 😁. It is possible to use a deployment with `files` for your use case? I'm not 100% clear on the need for repo/commit IDs (could be missing something here though).

Reply:
we need to deploy from a git repo, so that vercel clones the repo and deploy it on their side. Using `files` I need to have the files locally, we are not deploying `example_app` from the `example_app` git repo, but rather we are deploying `example_app1`, `example_app2`, `example_app3` ... from a `devops_terraform` git repo

Reply:
Ahh okay, I understand. It's a pretty unusual workflow but I guess it makes sense. :) I will try to prioritise getting this in place. It's just tricky supporting the different git providers.

Reply:
Out of curiosity have you considered putting them in a single repository, behind something like https://turborepo.org/ ?

Reply:
we're actually moving on from a monorepo to a distributed system to allow each team to manage their own code as the teams get bigger

Reply:
Makes sense! As an update on this issue - I've been having a play around, and I'm coming to the belief that the only thing needed is a `ref` field on the `vercel_deployment` resource. 

For example 
```terraform
resource "vercel_project" "example" {
   name = "my-test-project" 
   git_repository = {
       type = "github" 
       repo = "acme/example"
   }
}

# Deploy the main branch of acme/example. 
resource "vercel_deployment" "example" {
    project_id = vercel_project.example.id
    ref        = "main" # or a commit ID. 
}
```

The deployment resource can, itself, look up the relevant repoId / equivalent bitbucket/gitlab parameters, and pass these onto the deployment creation. 

You could even use it with a data-source to pull a project by a name: 
```terraform 
data "vercel_project" "example" {
    name = "my-test-project" 
}

resource "vercel_deployment" "example" {
    project_id = data.vercel_project.example.id
    ref        = "main" 
}
```

We already have a hard requirement that for a deployment to be created, a project must already exist. I think using the git-connection this way would work nicely, and also prevents any confusing fields being presented back out of the provider (like repoId or projectUuid or workspaceUuid). What do you think @WCMinor? Would this work for your use-case? 

With my playing around I think i'm fairly close to a working version, so if this works for you, I should be able to have a PR up soonish.

Reply:
This looks great yeah, much easier to understand and to configure.

Funny enough, if you pass a public github repo (not configured in your project) to the deploy API , it works, it deploys a project based on such public repo. That is the only case in which it would make sense to use a standalone `git_source` parameter. Honestly, I think that functionality is kinda out of the scope of this provider as it is not really documented by Vercel.

Reply:
Yeah. It's probably more a result of accidental API design than anything else 😓